### PR TITLE
search: introduce basic limiter

### PR DESCRIPTION
**Pull request by stefanhengl**
This adds a limiter to our GraphQL API that limits requests from
anonymous callers. The rate limit is configurable through the
site configuration.

There was already a much more sophisticated, but disabled,
rate limiter in place that currently undergoes testing. This new
limiter is a heavily simplified version of the existing limiter which we
can use for the time being.

To abstract from the implementation, I created 2 new interfaces
"Limiter" and "LimitWatcher".

### CODE REVIEW

File: schema/site.schema.json [hunk 1]
```diff
@@ -49,6 +49,11 @@
       "type": "object",
       "additionalProperties": false,
       "properties": {
+        "rateLimitAnonymous": {
+          "description": "Configures the hourly rate limits for anonymous calls to the GraphQL API. Setting limit to 0 disables the limiter. This is only relevant if unauthenticated calls to the API are permitted.",
+          "type": "integer",
+          "default": 100
```
Comment (ryanslade) [inline schema/site.schema.json:55]: This seems like a pretty low default.

**Review by ryanslade — COMMENTED** [review]
LGTM

There's probably a smaller change that would involve modifying the existing limiter but this also looks fine for now as we won't have time to look into the other rate limiter for at least a month.

File: schema/site.schema.json (hunk 1 above)
Comment (ryanslade) [reply schema/site.schema.json:55]: This a per hour for ALL anonymous requests right?

Comment (stefanhengl) [reply schema/site.schema.json:55]: This limits unnamed, anonymous queries not coming from the browser. Looking at the logs on honeycomb we get around 50 queries an hour that fall into that category during normal operations. We can start with something higher and then lower it if we need to.

Comment (ryanslade) [reply schema/site.schema.json:55]: 👍

**Review by keegancsmith — APPROVED** [review]
Did you look into legitimate unnamed anon requests? When we are getting attacked we don't want to fail those? Or do we have no other choice?

**Comment by stefanhengl** [comment]
The parametrized GQL queries match the format of our src-cli, so either it is anonymous traffic coming from src-cli or from something that just copied the query. Either way, it seems ok to limit those.

I couldn't match the other queries, EG
```
query { repository(name: "github.com/0ctobot/neutrino_kernel_oneplus_sdm845") { defaultBranch { target { commit { blob(path: "net/ipv4/tcp_ipv4.c") { content } } } } } }
```
They don't seem to come from us, but they could come from an extension?

**Comment by keegancsmith** [comment]
> The parametrized GQL queries match the format of our src-cli, so either it is anonymous traffic coming from src-cli or from something that just copied the query. Either way, it seems ok to limit those.

Agreed.

> I couldn't match the other queries, EG
> 
> ```
> query { repository(name: "github.com/0ctobot/neutrino_kernel_oneplus_sdm845") { defaultBranch { target { commit { blob(path: "net/ipv4/tcp_ipv4.c") { content } } } } } }
> ```
> 
> They don't seem to come from us, but they could come from an extension?

The repos it is running on when we looked at it is very curious to me. Too many random small repos. Which makes me think it isn't from one of our extensions, otherwise we should be seeing traffic on popular repos as well.

**Comment by bobheadxi** [comment]
@stefanhengl is there any progress on the rate limiter here? I stumbled upon the disabled rate limiter and was a bit confused as to why it exists when it's only used in tests

**Comment by stefanhengl** [comment]
> @stefanhengl is there any progress on the rate limiter here?

The basic rate limiter from this PR was introduced to fight an attacker from last year. I believe the disabled limiter was more sophisticated and tried to estimate things like a "query cost". I am not aware that any work on rate limiting happened since then.

**Pull request merged**